test/12_dependency_versions/cibuildwheel_test.py

```diff
@@ -0,0 +1,115 @@
+import os
+import re
+import pytest
+import textwrap
+import cibuildwheel.util
+
+import utils
+
+VERSION_REGEX = r'([\w-]+)==([^\s]+)'
+
+
+def get_versions_from_constraint_file(constraint_file):
+    with open(constraint_file, encoding='utf8') as f:
+        constraint_file_text = f.read()
+
+    versions = {}
+
+    for package, version in re.findall(VERSION_REGEX, constraint_file_text):
+        versions[package] = version
+
+    return versions
+
+
+@pytest.mark.parametrize('python_version', ['2.7', '3.x'])
+def test_pinned_versions(python_version):
+    if utils.platform == 'linux':
+        pytest.skip('linux doesn\'t pin individual tool versions, it pins manylinux images instead')
+
+    project_dir = os.path.dirname(__file__)
+
+    build_environment = {}
+
+    if python_version == '2.7':
+        constraint_filename = 'constraints-python27.txt'
+    else:
+        constraint_filename = 'constraints.txt'
+
+    constraint_file = os.path.join(cibuildwheel.util.resources_dir, constraint_filename)
+    constraint_versions = get_versions_from_constraint_file(constraint_file)
+
+    for package in ['pip', 'setuptools', 'wheel', 'virtualenv']:
+        env_name = 'EXPECTED_{}_VERSION'.format(package.upper())
+        build_environment[env_name] = constraint_versions[package]
+
+    cibw_environment_option = ' '.join(
+        ['{}={}'.format(k, v) for k, v in build_environment.items()]
+    )
+
+    # build and test the wheels
+    actual_wheels = utils.cibuildwheel_run(project_dir, add_env={
+        'CIBW_BUILD': '[cp]p27-*' if python_version == '2.7' else '[cp]p3?-*',
+        'CIBW_ENVIRONMENT': cibw_environment_option,
+    })
+
+    # also check that we got the right wheels
+    if python_version == '2.7':
+        expected_wheels = [w for w in utils.expected_wheels('spam', '0.1.0')
+                           if '-cp27' in w or '-pp27' in w]
+    else:
+        expected_wheels = [w for w in utils.expected_wheels('spam', '0.1.0')
+                           if '-cp27' not in w and '-pp27' not in w]
+
+    assert set(actual_wheels) == set(expected_wheels)
+
+
+@pytest.mark.parametrize('python_version', ['2.7', '3.x'])
+def test_dependency_constraints_file(tmp_path, python_version):
+    if utils.platform == 'linux':
+        pytest.skip('linux doesn\'t pin individual tool versions, it pins manylinux images instead')
```

Comment:
maybe change order of test. This first and `test_pinned_versions` second? This will force to perform installation package. 

Otherwise in `test_pinned_versions` there is no installation steep, because everything is already installed.

Reply:
good idea 👍

Reply:
Not sure I completely understand, but if you make a change that makes order matter, let's have a comment as warning there!

Reply:
Decided against this in the end, we don't really care if an installation took place as long as the versions are correct.